File: Solitaire.Api/Mappers/GameMapper.cs
using Microsoft.FSharp.Collections;
using Solitaire.Api.Models;
using System;
using System.Linq;
using static MoveHandler;

namespace Solitaire.Api.Mappers
{
    public class GameMapper : IGameMapper
    {
        private CardSuitWeb GetCardSuitWebFromCard(Game.Suit suit)
        {
            if (suit == Game.Suit.Hearts)
            {
                return CardSuitWeb.CLUBS;
            }
            if (suit.IsDiamonds)
            {
                return CardSuitWeb.DIAMONDS;
            }
            if (suit.IsSpades)
            {
                return CardSuitWeb.SPADES;
            }
            if (suit.IsHearts)
            {
                return CardSuitWeb.HEARTS;
            }
            return CardSuitWeb.HEARTS;
        }

        private Game.Suit MapCardSuitWebToCardSuit(CardSuitWeb cardSuiteWeb)
        {
            switch (cardSuiteWeb)
            {
                case CardSuitWeb.DIAMONDS:
                    return Game.Suit.Diamonds;
                case CardSuitWeb.CLUBS:
                    return Game.Suit.Clubs;
                case CardSuitWeb.SPADES:
                    return Game.Suit.Spades;
                case CardSuitWeb.HEARTS:
                    return Game.Suit.Hearts;
                default:
                    return Game.Suit.Hearts;
            }
        }

        private CardWeb MapCardToCardWeb(Game.Card card)
        {
            return new CardWeb
            {
                Rank = card.Rank,
                Suit = GetCardSuitWebFromCard(card.Suit)
            };
        }

        private Game.Card MapCardWebToCard(CardWeb cardWeb)
        {
            return new Game.Card(cardWeb.Rank, MapCardSuitWebToCardSuit(cardWeb.Suit));
        }

        public GameWeb MapGameToGameWeb(Game.Game game)
        {
            return new GameWeb
            {
                Stock = game.Stock.ToList().Select(MapCardToCardWeb).ToList(),
                Wastepile = game.Wastepile.ToList().Select(MapCardToCardWeb).ToList(),
                Active = game.ActiveStock.ToList().Select(MapCardToCardWeb).ToList(),
                Foundations = game.Foundations.Select((column) => column.Select(MapCardToCardWeb).ToList()).ToList(),
                Tableau = game.Tableau.Select((column) => column.Select(MapCardToCardWeb).ToList()).ToList(),
            };
        }

        public Game.Game MapGameWebToGame(GameWeb gameWeb)
        {
            var stock = ListModule.OfSeq(gameWeb.Stock.Select(MapCardWebToCard));
            var active = ListModule.OfSeq(gameWeb.Active.Select(MapCardWebToCard));
            var wastepile = ListModule.OfSeq(gameWeb.Wastepile.Select(MapCardWebToCard));

            var tableau = ListModule.OfSeq(gameWeb.Tableau.Select(
                (column) => ListModule.OfSeq(column.Select(MapCardWebToCard))
            ));

            var foundations = ListModule.OfSeq(gameWeb.Tableau.Select(
                (column) => ListModule.OfSeq(column.Select(MapCardWebToCard))
             ));

            return new Game.Game(tableau, stock, active, wastepile, foundations);
        }

        public CardArea MapCardAreaWebToCardGame(CardAreaWeb cardAreaWeb)
        {
            return cardAreaWeb.CardArea switch
            {
                CardAreaTypeWeb.Tableau => CardArea.NewTableau(cardAreaWeb.Index),
                CardAreaTypeWeb.Foundation => CardArea.NewTableau(cardAreaWeb.Index),
                CardAreaTypeWeb.Active => CardArea.ActiveStock,
                _ => throw new ArgumentException("Invalid card area specified")
            };
        }
    }
}

File: Solitaire.WebApi/Mappers/GameMapper.fs
module GameMapper

open Game
open GameWeb
open System

[<Literal>]
let ClubsSuit = "CLUBS"

[<Literal>]
let DiamondsSuit = "DIAMONDS"

[<Literal>]
let HeartsSuit = "HEARTS"

[<Literal>]
let SpadesSuit = "SPADES"

let private mapCardToCardWeb (card: Card) : CardWeb =
    let suit =
        match card.Suit with
        | Clubs -> ClubsSuit
        | Hearts -> HeartsSuit
        | Spades -> SpadesSuit
        | Diamonds -> DiamondsSuit

    { Suit = suit; Rank = card.Rank }

let private mapCardWebToCard (card: CardWeb) : Card =
    let suit =
        match card.Suit with
        | ClubsSuit -> Clubs
        | DiamondsSuit -> Diamonds
        | HeartsSuit -> Hearts
        | SpadesSuit -> Spades
        | _ -> raise (ArgumentException("Attempting to map invalid suit value"))

    { Suit = suit; Rank = card.Rank }

let private mapCardListToCardListWeb (cards: Card list) : CardWeb list = cards |> List.map (mapCardToCardWeb)

let private map2dCardListTo2dCardListWeb (cards: Card list list) : CardWeb list list =
    cards |> List.map (mapCardListToCardListWeb)

let private mapCardListWebToCardList (cards: CardWeb list) : Card list = cards |> List.map (mapCardWebToCard)

let private map2dCardListWebTo2dCardList (cards: CardWeb list list) : Card list list =
    cards |> List.map (mapCardListWebToCardList)

let mapGameToGameWeb (game: Game) : GameWeb =
    { Tableau = map2dCardListTo2dCardListWeb (game.Tableau)
      Foundations = map2dCardListTo2dCardListWeb (game.Foundations)
      ActiveStock = mapCardListToCardListWeb (game.ActiveStock)
      Stock = mapCardListToCardListWeb (game.Stock)
      Wastepile = mapCardListToCardListWeb (game.Wastepile) }

let mapGameWebToGame (game: GameWeb) : Game =
    { Tableau = map2dCardListWebTo2dCardList (game.Tableau)
      Foundations = map2dCardListWebTo2dCardList (game.Foundations)
      ActiveStock = mapCardListWebToCardList (game.ActiveStock)
      Stock = mapCardListWebToCardList (game.Stock)
      Wastepile = mapCardListWebToCardList (game.Wastepile) }
